src/keys/bip44.rs

```diff
@@ -0,0 +1,219 @@
+        S: Segment + TryFrom<u32>,
+        <S as TryFrom<u32>>::Error: core::fmt::Debug,
+    {
+        let chain: [_; 5] = self.to_chain::<K>();
+
+        // maximum number segments is 2^31, trim usize value to fit u32
+        let address_count = core::cmp::min(1 << 31, address_count) as u32;
+
+        // BIP44 conversion rules are strict, the last element is address_index
+        let address_start = chain[4];
+        let hardening_bit: u32 = address_start.into() & slip10::HARDEN_MASK;
+        // strip hardening bit as it may interfere and overflow
+        let unhardened_start: u32 = address_start.unharden().into();
+        // this is guaranteed to not overflow and be <= 2^31
+        let unhardened_end: u32 = core::cmp::min(1_u32 << 31, unhardened_start + address_count);
+
+        // this is the final range guaranteed to not overflow address_index space
+        let child_segments = (unhardened_start..unhardened_end).map(move |unhardened_address_index| -> S {
+            let address_index = hardening_bit | unhardened_address_index;
+            // SAFETY: address_index is guaranteed to have the correct hardening as the target type `S`, so unwrap()
+            // can't fail
+            address_index.try_into().unwrap()
+        });
+
+        let mk = if child_segments.len() > 0 {
+            m.derive(chain[..4].iter().copied())
+        } else {
+            // no need to derive mk if there's no child_segments, just use empty/zero one
+            slip10::Slip10::new()
+        };
+        mk.children(child_segments)
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct BadPurpose;
+
+impl From<BadPurpose> for crate::Error {
+    fn from(inner: BadPurpose) -> Self {
+        crate::Error::Bip44Error(inner)
+    }
+}
+
+impl TryFrom<[u32; 5]> for Bip44 {
+    type Error = BadPurpose;
+    fn try_from(segments: [u32; 5]) -> Result<Self, Self::Error> {
+        if let [Bip44::PURPOSE, coin_type, account, change, address_index] = segments {
+            Ok(Self {
+                coin_type,
+                account,
+                change,
+                address_index,
+            })
+        } else {
+            Err(BadPurpose)
+        }
+    }
+}
+
+impl From<[u32; 4]> for Bip44 {
+    fn from(segments: [u32; 4]) -> Self {
+        let [coin_type, account, change, address_index] = segments;
+        Self {
+            coin_type,
+            account,
+            change,
+            address_index,
+        }
+    }
+}
+
+impl From<&Bip44> for [u32; 5] {
+    fn from(bip44_chain: &Bip44) -> [u32; 5] {
+        [
+            Bip44::PURPOSE,
+            bip44_chain.coin_type,
+            bip44_chain.account,
+            bip44_chain.change,
+            bip44_chain.address_index,
+        ]
+    }
+}
+
+pub struct Bip44Builder(Bip44);
+
+impl Bip44Builder {
+    pub fn new() -> Self {
```

Comment:
I would probably replace this with `#[derive(Default)]`

Reply:
Removed Builder. Moved setters into Bip44. Derive Default.